src/matchmaker.ts

```diff
@@ -1,44 +1,91 @@
 import { db } from "@siggame/colisee-lib";
+import * as knex from "knex";
 import * as _ from "lodash";
 import * as winston from "winston";
 
+import { generateDerangedPairs, permute } from "./helpers";
 import * as vars from "./vars";
 
-export class Matchmaker {
+interface IRecentSub {
+    id: number;
+    teamId: number;
+}
+
+interface IMatchmakerOptions {
+    interval: number;
+    matchReplications: number;
+    maxMatches: number;
+}
+export class Matchmaker implements IMatchmakerOptions {
 
-    private intervalId: any;
+    private intervalId?: NodeJS.Timer;
 
-    constructor() {
+    constructor(
+        public interval: number = vars.SCHED_INTERVAL,
+        public matchReplications: number = vars.REPLICATIONS,
+        public maxMatches: number = vars.SCHED_MAX) {
         this.intervalId = undefined;
     }
 
     start() {
-        this.intervalId = setInterval(() => { this.poll(); }, vars.SCHED_INTERVAL);
+        if (_.isNil(this.intervalId)) {
+            this.intervalId = setInterval(() => {
+                this.poll().catch((e) => { winston.error(e); });
+            }, vars.SCHED_INTERVAL);
+        }
     }
 
     stop() {
-        clearInterval(this.intervalId);
+        if (!_.isNil(this.intervalId)) {
+            clearInterval(this.intervalId);
+            this.intervalId = undefined;
+        }
     }
 
-    async getTeamsRandomOrder(): Promise<db.Team[]> {
-        return db.connection("teams").select("*")
-            .then(db.rowsToTeams)
-            .then((teams) => teams) // add permutation
+    async getPairedTeams() {
+        return db.connection.from(function (this: knex.QueryBuilder) {
+            this.from("submissions as subs")
+                .select("subs.id as subId", "subs.team_id as teamId", db.connection.raw("max(subs.version) as recent_version"))
+                .where({ status: "finished" })
+                .groupBy("subId")
+                .as("recent_subs");
+        }).join("submissions", function () {
+            this.on("recent_subs.teamId", "submissions.team_id").andOn("recent_subs.recent_version", "submissions.version");
+        })
+            .select("submissions.id as id", "recent_subs.teamId")
+            .then(permute)
+            .then(generateDerangedPairs)
+            .then((pairs): [IRecentSub, IRecentSub][] => pairs)
             .catch((e: Error) => { throw e; });
     }
 
     async scheduledNum(): Promise<number> {
-        return 4;
+        return db.connection("games").where({ status: "queued" }).count("*")
+            .then(([{ count }]): number => count)
+            .catch((e: Error) => { throw e; });
     }
 
     async poll(): Promise<void> {
-        let teams: db.Team[] = [];
-        try {
-            teams = await this.getTeamsRandomOrder();
-        } catch (e) {
-            winston.error(e);
-        }
+        if (await this.scheduledNum() < vars.SCHED_MAX) {
```

Comment:
Nothing to change here; just an idea.
Capture the scheduled num variable and log it if its above some threshold as a critical error.

For example; Set the threshold to 200, so that if we have more than 1000 games queued, we should probably be alerted that something is wrong.